presto-hash/src/main/java/com/facebook/presto/hash/ConsistentHashingNodeProvider.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hash;
+
+import com.facebook.presto.spi.NodeProvider;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static com.google.common.hash.Hashing.md5;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.unmodifiableList;
+
+public class ConsistentHashingNodeProvider<T>
+        implements NodeProvider<T>
+{
+    private final SortedMap<Integer, T> circle = new TreeMap<>();
+
+    public ConsistentHashingNodeProvider(Collection<T> nodes)
+    {
+        for (T node : nodes) {
+            add(node);
+        }
+    }
+
+    /**
+     * Get target nodes for a particular key
+     * By first locating the node for the key and clockwise append the rest of the nodes in hash ring to the target list
+     * Until we have reached the number of nodes or we have exhausted the hash ring
+     * @param key
+     * @param count specifies how many nodes to look for
+     * @return
+     */
+    public List<T> get(Object key, int count)
+    {
+        if (circle.isEmpty() || key == null) {
+            return unmodifiableList(new ArrayList<>());
+        }
+
+        List<T> nodes = new ArrayList<>();
+        nodes.add(get(key).get());
+        while (count > 1 && nodes.size() < circle.size()) {
+            nodes.add(get(key.hashCode() + 1).get());
+            count--;
+        }
+        return unmodifiableList(nodes);
+    }
+
+    /**
+     * Get the target node for a particular key
+     * @param key
+     * @return
+     */
+    public Optional<T> get(Object key)
+    {
+        if (circle.isEmpty() || key == null) {
+            return Optional.empty();
+        }
+        int hash = md5().hashString(key.toString(), UTF_8).asInt();
```

Comment:
why not just use `key.hashCode()` directly? The reason alluxio uses md5 is because it has to have 0 conflict. But it's not the case here. Then you can move everything to SPI.

Reply:
The reason I'm using md5 is that hash function has much higher possibility of collision, and when building our hash ring, we want to avoid that, thus for `add` method, we have to use md5.
Here we can directly use  `key.hashCode()`, but given that we already introduced md5, we might as well take advantage of it.

Reply:
Md5 is completely downgraded here. It's not because md5 has less hash collision. It is because you use `int`. That is a 32-bit hash space. Not matter what hash function you use. That is going to cause high collision. Just remove md5 and move this class to SPI.